### PR TITLE
KAFKA-14628: Move CommandLineUtils and CommandDefaultOptions shared classes

### PR DESCRIPTION
**Pull request by fvaleri**
`CommandLineUtils` and `CommandDefaultOptions` are required by most commands, so they must be migrated first. This PR can be used as the base for other commands migration.

### CODE REVIEW

**Comment by fvaleri** [comment]
@mimaison @ijuma

**Comment by fvaleri** [comment]
@clolov @vamossagar12 @tinaselenge

**Comment by fvaleri** [comment]
Take a look at DumpLogSegments to see how CommandDefaultOptions is used. It is a much better approach than building the script option list at the start of the main/execute method. In Java, you simply create a private static class that extends CommandDefaultOptions and put all parsing logic there.

File: server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java
Comment (clolov) [inline server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java:137]: Do we need the `Exit.exit(1)` introduced elsewhere in this pull request if this method already uses it internally?

Comment (fvaleri) [reply server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java:137]: Internally? WDYM?

There are some places, where I call this method and then call `Exit.exit(1)` (the Scala version). This is because I need `Nothing` as return value and there is no Java equivalent. It's an ugly hack, but it will naturally go away with the migration of these commands.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala
Comment (vamossagar12) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: The addition of `OptionSpec` and `withOptionalArg` is not necessary strictly right?

File: server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java
Comment (vamossagar12) [reply server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java:137]: I had the same question as @clolov , but I guess there's no getting around what you are saying. Either ways, would it make sense to consolidate all the duplicate method calls i.e

```
CommandLineUtils.printUsageAndDie(opts.parser, message)
Exit.exit(1)
```

into a separate method and add a comment to it saying the scala Exit is temporary and not needed when migrating to Java? That should make it easier for the immediate future migrations and reduce the duplicacy. WDYT?

File: core/src/main/scala/kafka/admin/ConfigCommand.scala
Comment (fvaleri) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: Correct, it is a left over from a previous attempt. Reverting.

**Comment by fvaleri** [comment]
@clolov @vamossagar12 took your suggestions. Thanks.

File: server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java
Comment (fvaleri) [reply server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java:137]: Yeah, good idea. Done.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala
Comment (vamossagar12) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: Thanks @fvaleri . Wondering if these and other similar lines can be changed to use ToolsUtils. printUsageAndDie()?

File: core/src/main/scala/kafka/utils/ToolsUtils.scala
Comment (vamossagar12) [inline core/src/main/scala/kafka/utils/ToolsUtils.scala]: I am thinking if we can add another parameter to this to say if we want to do an Exit(1) from this method or not. I see some places where Exit(1) is not being invoked from the the tools classes. WDYT?

**Comment by vamossagar12** [comment]
> @clolov @vamossagar12 took your suggestions. Thanks.

Thanks @fvaleri ! Couple of minor comments. This should be good to go after that!

File: core/src/main/scala/kafka/utils/ToolsUtils.scala
Comment (fvaleri) [reply core/src/main/scala/kafka/utils/ToolsUtils.scala]: In the original code we always call exit, as the method name implies, so I think it's fine.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala
Comment (fvaleri) [reply core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: Yes, there were some some replacements that I missed. Fixed.

**Comment by vamossagar12** [comment]
Thanks @fvaleri . LGTM.

File: server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java
Comment (clolov) [reply server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java:137]: ~~Ah, got it. I wasn't aware that Scala needs a different Exit to the one already defined in this method! I agree with the subsequent suggestion.~~

File: core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala
Comment (clolov) [inline core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]: Could you use the new `ToolsUtils.printUsageAndDie(...)` which you introduced?

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (clolov) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: A similar question about using `ToolsUtils.printUsageAndDie(...)` here as well.

File: server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java
Comment (clolov) [reply server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java:137]: Actually, sorry, I still don't get it and I should have explained my question better. What I meant to ask is whether we can put all calls to `Exit.exit(1)` only in `CommandLineUtils` functions where we expect a termination of the program and not make explicit reference to it in other commands? For example, here https://github.com/apache/kafka/blob/0803659dc5b5e0aeae82b29e6ba4d07cb855bc1c/core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala#L74 we have pushed `Exit.exit(1)` down into `CommandLineUtils`. And here https://github.com/apache/kafka/blob/0803659dc5b5e0aeae82b29e6ba4d07cb855bc1c/core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala#L103-L104 we have not.

File: core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala
Comment (fvaleri) [reply core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]: Sure. Thanks.

**Review by clolov — COMMENTED** [review]
> Actually, sorry, I still don't get it and I should have explained my question better. What I meant to ask is whether we can put all calls to `Exit.exit(1)` only in `CommandLineUtils` functions where we expect a termination of the program and not make explicit reference to it in other commands? For example, here
> 
> https://github.com/apache/kafka/blob/0803659dc5b5e0aeae82b29e6ba4d07cb855bc1c/core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala#L74
> we have pushed `Exit.exit(1)` down into `CommandLineUtils`. And here
> 
> https://github.com/apache/kafka/blob/0803659dc5b5e0aeae82b29e6ba4d07cb855bc1c/core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala#L103-L104
> we have not.

Reposting my question here as well, because I thought that continuing a resolved discussion will unresolve it, but apparently it does not :(

File: server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java
Comment (mimaison) [inline server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java]: Should we take this opportunity to update this name?

File: server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java
Comment (mimaison) [inline server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java]: I wonder if `printUsageAndExit` might be a more appropriate name. Same below. WDYT?

File: server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java
Comment (mimaison) [inline server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java]: The `null` check should not be needed here.

**Review by mimaison — COMMENTED** [review]
Thanks for the PR, it looks good overall. I left a few suggestions.

File: server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java
Comment (fvaleri) [reply server-common/src/main/java/org/apache/kafka/server/util/CommandLineUtils.java]: What about maybePrintHelpOrVersion? 

I think the exit part is kind of implicit when you ask for help or version.

**Comment by fvaleri** [comment]
> we have pushed Exit.exit(1) down into CommandLineUtils.

@clolov it wasn't pushed down, that's the original logic. The code you are referring to in ZkSecurityMigrator.scala:103 is outdated. Now we are using ToolsUtils.printUsageAndExit in a few places for the reason expressed in the method comment. Please, pull latest changes and let me know.

**Review by mimaison — APPROVED** [review]
LGTM, I'll wait for the CI to complete and then merge

**Comment by ijuma** [comment]
Thanks for doing this. One thing I didn't understand is why the shared classes are in `server-common` instead of `tools`. Is this because `core` also uses the class?

**Comment by fvaleri** [comment]
> Thanks for doing this. One thing I didn't understand is why the shared classes are in `server-common` instead of `tools`. Is this because `core` also uses the class?

Good question. CommandLineUtils is mostly used by tools, but Kafka.scala has a dependency on it. In order to move these 2 shared classes to the tools module, we would need to add tools dependency to core. Is this correct?

**Comment by ijuma** [comment]
Got it, so it's useful for `main` methods too - not just CLI tools. Then it's fine for it to be in `server-common`.

**Pull request merged**
